File: epms/page/epms_dashboard/epms_dashboard.js
// Copyright (c) 2024, EPMS Team and contributors
// For license information, please see license.txt

frappe.pages['epms-dashboard'].on_page_load = function(wrapper) {
    var page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __('Employee Performance Dashboard'),
        single_column: true
    });

    // Add custom buttons
    page.add_button(__('Refresh'), function() {
        load_dashboard();
    }, 'fa fa-refresh');

    page.add_button(__('Export Report'), function() {
        frappe.set_route('query-report', 'Monthly Performance Report');
    }, 'fa fa-download');

    // Load dashboard
    load_dashboard();
};

function load_dashboard() {
    frappe.call({
        method: 'epms.epms.page.epms_dashboard.epms_dashboard.get_dashboard_data',
        callback: function(r) {
            if (r.message) {
                render_dashboard(r.message);
            }
        }
    });
}

function render_dashboard(data) {
    var wrapper = $('#page-epms-dashboard .page-content');
    wrapper.empty();

    // Determine user role
    var isFounder = frappe.boot.epms_is_founder;
    var isTeamLeader = frappe.boot.epms_is_team_leader;
    var isTeamMember = frappe.boot.epms_is_team_member;

    var html = '<div class="row">';

    if (isFounder) {
        html += render_founder_dashboard(data);
    } else if (isTeamLeader) {
        html += render_team_leader_dashboard(data);
    } else if (isTeamMember) {
        html += render_employee_dashboard(data);
    }

    html += '</div>';
    wrapper.html(html);

    // Initialize charts
    init_charts();
}

function render_founder_dashboard(data) {
    var html = '';

    // Stats Cards
    html += '<div class="col-md-3"><div class="card text-white bg-primary mb-3"><div class="card-body"><h5 class="card-title">Total Employees</h5><h2 class="card-text">' + (data.total_employees || 0) + '</h2></div></div></div>';
    html += '<div class="col-md-3"><div class="card text-white bg-success mb-3"><div class="card-body"><h5 class="card-title">Total Teams</h5><h2 class="card-text">' + (data.total_teams || 0) + '</h2></div></div></div>';
    html += '<div class="col-md-3"><div class="card text-white bg-info mb-3"><div class="card-body"><h5 class="card-title">Tasks Today</h5><h2 class="card-text">' + (data.tasks_today || 0) + '</h2></div></div></div>';
    html += '<div class="col-md-3"><div class="card text-white bg-warning mb-3"><div class="card-body"><h5 class="card-title">Pending Tasks</h5><h2 class="card-text">' + (data.pending_tasks || 0) + '</h2></div></div></div>';

    html += '<div class="row">';
    html += '<div class="col-md-3"><div class="card text-white bg-danger mb-3"><div class="card-body"><h5 class="card-title">Blocked Tasks</h5><h2 class="card-text">' + (data.blocked_tasks || 0) + '</h2></div></div></div>';
    html += '<div class="col-md-3"><div class="card text-white bg-secondary mb-3"><div class="card-body"><h5 class="card-title">Avg Performance</h5><h2 class="card-text">' + (data.avg_performance || 0) + '%</h2></div></div></div>';

    if (data.top_performer) {
        html += '<div class="col-md-3"><div class="card text-white bg-success mb-3"><div class="card-body"><h5 class="card-title">Top Performer</h5><p class="card-text">' + data.top_performer.employee_name + '</p><small>' + data.top_performer.overall_score + '</small></div></div></div>';
    }

    if (data.lowest_performer) {
        html += '<div class="col-md-3"><div class="card text-white bg-danger mb-3"><div class="card-body"><h5 class="card-title">Needs Improvement</h5><p class="card-text">' + data.lowest_performer.employee_name + '</p><small>' + data.lowest_performer.overall_score + '</small></div></div></div>';
    }

    html += '</div>';

    // Charts section
    html += '<div class="row"><div class="col-md-6"><div class="card mb-3"><div class="card-header">Performance Distribution</div><div class="card-body"><div id="performance-distribution-chart"></div></div></div></div>';
    html += '<div class="col-md-6"><div class="card mb-3"><div class="card-header">Monthly Trend</div><div class="card-body"><div id="monthly-trend-chart"></div></div></div></div></div>';

    return html;
}

function render_team_leader_dashboard(data) {
    var html = '';

    html += '<div class="col-md-3"><div class="card text-white bg-primary mb-3"><div class="card-body"><h5 class="card-title">Today\'s Tasks</h5><h2 class="card-text">' + (data.today_tasks || 0) + '</h2></div></div></div>';
    html += '<div class="col-md-3"><div class="card text-white bg-warning mb-3"><div class="card-body"><h5 class="card-title">Pending Tasks</h5><h2 class="card-text">' + (data.pending_tasks || 0) + '</h2></div></div></div>';
    html += '<div class="col-md-3"><div class="card text-white bg-success mb-3"><div class="card-body"><h5 class="card-title">Completed Tasks</h5><h2 class="card-text">' + (data.completed_tasks || 0) + '</h2></div></div></div>';
    html += '<div class="col-md-3"><div class="card text-white bg-info mb-3"><div class="card-body"><h5 class="card-title">Team Score</h5><h2 class="card-text">' + (data.team_score || 0) + '</h2></div></div></div>';

    html += '</div>';

    // Charts section
    html += '<div class="row"><div class="col-md-6"><div class="card mb-3"><div class="card-header">Employee Productivity</div><div class="card-body"><div id="employee-productivity-chart"></div></div></div></div>';
    html += '<div class="col-md-6"><div class="card mb-3"><div class="card-header">Task Status</div><div class="card-body"><div id="task-status-chart"></div></div></div></div></div>';

    return html;
}

function render_employee_dashboard(data) {
    var html = '';

    if (data.today_performance && data.today_performance.length > 0) {
        var today = data.today_performance[0];
        html += '<div class="col-md-3"><div class="card text-white bg-primary mb-3"><div class="card-body"><h5 class="card-title">Today\'s Rating</h5><h2 class="card-text">' + (today.daily_rating || 'N/A') + '</h2></div></div></div>';
    } else {
        html += '<div class="col-md-3"><div class="card text-white bg-secondary mb-3"><div class="card-body"><h5 class="card-title">Today\'s Rating</h5><h2 class="card-text">No Entry</h2></div></div></div>';
    }

    if (data.monthly_score) {
        html += '<div class="col-md-3"><div class="card text-white bg-success mb-3"><div class="card-body"><h5 class="card-title">Monthly Score</h5><h2 class="card-text">' + (data.monthly_score.overall_score || 0) + '</h2></div></div></div>';
        html += '<div class="col-md-3"><div class="card text-white bg-info mb-3"><div class="card-body"><h5 class="card-title">Completed Tasks</h5><h2 class="card-text">' + (data.monthly_score.tasks_completed || 0) + '</h2></div></div></div>';
        html += '<div class="col-md-3"><div class="card text-white bg-warning mb-3"><div class="card-body"><h5 class="card-title">Overall Grade</h5><h2 class="card-text">' + (data.monthly_score.final_grade || 'N/A') + '</h2></div></div></div>';
    }

    html += '</div>';

    // Charts section
    html += '<div class="row"><div class="col-md-6"><div class="card mb-3"><div class="card-header">Monthly Performance</div><div class="card-body"><div id="monthly-performance-chart"></div></div></div></div>';
    html += '<div class="col-md-6"><div class="card mb-3"><div class="card-header">Score Trend</div><div class="card-body"><div id="score-trend-chart"></div></div></div></div></div>';

    return html;
}

function init_charts() {
    // Initialize charts here
    // Charts will be loaded via frappe.chart
}

File: epms/employee_performance/page/epms_dashboard/epms_dashboard.py
import frappe
from frappe import _


@frappe.whitelist()
def get_dashboard_data():
    """Get dashboard data based on user role."""
    user_roles = frappe.get_roles(frappe.session.user)

    if "EPMS Founder" in user_roles:
        return get_founder_dashboard()
    elif "EPMS Team Leader" in user_roles:
        return get_team_leader_dashboard()
    elif "EPMS Team Member" in user_roles:
        return get_employee_dashboard()
    else:
        return {}


def get_founder_dashboard():
    """Get founder dashboard data."""
    from frappe.utils import getdate, nowdate

    today = getdate(nowdate())

    # Stats
    total_employees = frappe.db.count("User", {"user_type": "System User", "enabled": 1})
    total_teams = frappe.db.count("Team", {"status": "Active"})
    tasks_today = frappe.db.count(
        "Daily Performance", {"date": today, "docstatus": 1}
    )
    pending_tasks = frappe.db.count(
        "Pending Task",
        {"current_status": ["in", ["Pending", "In Progress"]], "docstatus": 1},
    )
    blocked_tasks = frappe.db.count(
        "Pending Task", {"current_status": "Blocked", "docstatus": 1}
    )

    # Average performance
    avg_performance = frappe.db.get_value(
        "Performance Scorecard",
        {"docstatus": 1},
        "avg(overall_score)",
    )

    # Top performer
    top_performer = frappe.get_all(
        "Performance Scorecard",
        {"docstatus": 1},
        ["employee_name", "overall_score"],
        order_by="overall_score desc",
        limit_page_length=1,
    )

    # Lowest performer
    lowest_performer = frappe.get_all(
        "Performance Scorecard",
        {"docstatus": 1},
        ["employee_name", "overall_score"],
        order_by="overall_score asc",
        limit_page_length=1,
    )

    return {
        "total_employees": total_employees,
        "total_teams": total_teams,
        "tasks_today": tasks_today,
        "pending_tasks": pending_tasks,
        "blocked_tasks": blocked_tasks,
        "avg_performance": round(avg_performance or 0, 2),
        "top_performer": top_performer[0] if top_performer else None,
        "lowest_performer": lowest_performer[0] if lowest_performer else None,
    }


def get_team_leader_dashboard():
    """Get team leader dashboard data."""
    from frappe.utils import getdate, nowdate

    today = getdate(nowdate())
    user = frappe.session.user

    # Get team
    team = frappe.db.get_value("Team", {"team_leader": user}, "name")

    if not team:
        return {}

    # Stats
    today_tasks = frappe.db.count(
        "Daily Performance", {"team": team, "date": today, "docstatus": 1}
    )
    pending_tasks = frappe.db.count(
        "Pending Task",
        {
            "employee": [
                "in",
                frappe.get_all(
                    "Team Member Mapping",
                    {"team": team, "status": "Active"},
                    pluck="user",
                ),
            ],
            "current_status": ["in", ["Pending", "In Progress"]],
            "docstatus": 1,
        },
    )
    completed_tasks = frappe.db.count(
        "Daily Performance",
        {"team": team, "task_status": "Completed", "docstatus": 1},
    )

    # Team average score
    current_month = today.month
    current_year = today.year
    team_score = frappe.db.get_value(
        "Performance Scorecard",
        {"team": team, "month": current_month, "year": current_year, "docstatus": 1},
        "avg(overall_score)",
    )

    # Top performer
    top_performer = frappe.get_all(
        "Performance Scorecard",
        {"team": team, "month": current_month, "year": current_year, "docstatus": 1},
        ["employee_name", "overall_score"],
        order_by="overall_score desc",
        limit_page_length=1,
    )

    # Lowest performer
    lowest_performer = frappe.get_all(
        "Performance Scorecard",
        {"team": team, "month": current_month, "year": current_year, "docstatus": 1},
        ["employee_name", "overall_score"],
        order_by="overall_score asc",
        limit_page_length=1,
    )

    return {
        "today_tasks": today_tasks,
        "pending_tasks": pending_tasks,
        "completed_tasks": completed_tasks,
        "team_score": round(team_score or 0, 2),
        "top_performer": top_performer[0] if top_performer else None,
        "lowest_performer": lowest_performer[0] if lowest_performer else None,
    }


def get_employee_dashboard():
    """Get employee dashboard data."""
    from frappe.utils import getdate, nowdate

    today = getdate(nowdate())
    user = frappe.session.user

    # Today's performance
    today_perf = frappe.get_all(
        "Daily Performance",
        {"employee": user, "date": today, "docstatus": 1},
        ["daily_rating", "task_status", "completion_percentage"],
    )

    # Monthly score
    current_month = today.month
    current_year = today.year
    scorecard = frappe.db.get_value(
        "Performance Scorecard",
        {"employee": user, "month": current_month, "year": current_year, "docstatus": 1},
        ["overall_score", "final_grade", "tasks_completed", "pending_tasks"],
        as_dict=True,
    )

    return {
        "today_performance": today_perf,
        "monthly_score": scorecard,
    }
